boot/freeldr/freeldr/arch/realmode/fathelp.inc

```diff
@@ -64,9 +72,19 @@ FatHelperEntryPoint:
     /* First save AX - the start cluster of freeldr.sys */
     push ax
 
+    /* Save ds */
+    push ds
+
+    /* Prepare ds before PutChars call */
+    mov bx, cs
+    mov ds, bx
+
     /* Display "Loading FreeLoader..." message */
-    mov si, offset msgLoading
-    call word ptr [bp-PutCharsOffset]
+    mov si, offset msgLoading 
+    call PutChars
```

Comment:
I wonder why it wouldn't be possible to just do a `call word ptr ds:[bp-PutCharsOffset]`
where `ds` (or whatever other segment) would be kept to the segment where the (FAT12/16 first) sector of the bootsector is still stored.
This would however probably require the bootsector code to patch the `ret` instruction of this PutChars function to a `retf` or something alike, just before jumping to freeldr execution, to take into account for the fact this would be far-calls that are now going to be done on these bootsector helpers.

Reply:
`call word ptr ss:[bp-PutCharsOffset]` only affects the instruction pointer.

Reply:
https://c9x.me/x86/html/file_module_x86_id_26.html
I'm looking at the "far call" thing.
> The far address is loaded directly into the CS and EIP registers.

Reply:
using "far call" thing will work but i would need to change the bootsector

Reply:
> `call word ptr ss:[bp-PutCharsOffset]` only affects the instruction pointer.

Ah yes indeed, we would actually need a segment:offset stored there ("PutCharsOffset") instead of just offset.

Reply:
> using "far call" thing will work but i would need to change the bootsector

My suggestion was so that if we could try to avoid the large `CallBootsectorFunction` stuff, it might be of good interest.

Maybe you could see why also I suggested the far-call stuff if you analyze the last ~ 10 instructions before the far jump 2000:0000 in the FAT12/16 bootsector of NT.

Reply:
i removed the large `CallBootsectorFunction` now.